warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java

```diff
@@ -0,0 +1,328 @@
+                // otherwise, add this node to the existing parent's list of children
+                if (!node.jjtGetParent().equals(parentStack.peek())) {
+                    parentStack.push(node.jjtGetParent());
+                    childrenStack.push(Lists.newArrayList(newNode));
+                } else {
+                    childrenStack.peek().add(newNode);
+                }
+            }
+        }
+        
+        return newNode;
+    }
+    
+    /**
+     * Iteratively creates a copy of the passed in JexlNode
+     *
+     * @param node
+     *            the node to be copied
+     * @param postOrderDeque
+     *            the post order traversal of the copied tree
+     * @return the copied tree
+     */
+    private JexlNode copyTree(JexlNode node, Deque<JexlNode> postOrderDeque) {
+        // add all the nodes to the stack and iterate...
+        Deque<JexlNode> workingStack = new LinkedList<>();
+        
+        // create a copy of this node which shares the same children as the original node
+        JexlNode copiedNode = rebuildNode(node);
+        workingStack.push(copiedNode);
+        
+        // compute the post order traversal of all of the nodes, and copy them
+        while (!workingStack.isEmpty()) {
+            JexlNode poppedNode = workingStack.pop();
+            postOrderDeque.push(poppedNode);
+            
+            // if this node has children, create copies of them
+            if (poppedNode.jjtGetNumChildren() > 0) {
+                List<JexlNode> copiedChildren = new ArrayList<>();
+                for (JexlNode child : children(poppedNode)) {
+                    if (child != null) {
+                        
+                        // create a copy of this node which shares the same children as the original node
+                        JexlNode copiedChild = rebuildNode(child);
+                        
+                        copiedChildren.add(copiedChild);
+                        workingStack.push(copiedChild);
+                    }
+                }
+                
+                // Reassign the children for this copied node
+                JexlNodes.children(poppedNode, copiedChildren.toArray(new JexlNode[copiedChildren.size()]));
+            }
+        }
+        
+        return copiedNode;
+    }
+    
+    /**
+     * Returns a copy of the passed in node.
+     *
+     * If the original node has children, those exact children (not copies) will be added to the copied node. However, the parentage of those child nodes will
+     * be left as-is.
+     *
+     * If the original node has no children, we will simply use the RebuildingVisitor to copy the node.
+     *
+     * @param node
+     *            the node to copy
+     * @return the copied node
+     */
+    private JexlNode rebuildNode(JexlNode node) {
+        JexlNode newNode;
+        if (node.jjtGetNumChildren() == 0) {
+            newNode = RebuildingVisitor.copy(node);
```

Comment:
Do we need this portion of the if?  I think the else part will work when numChildren == 0...